File: app/src/main/java/me/gm/cleaner/plugin/home/ui/HomeAdapter.java
/*
 * Copyright 2021 Green Mushroom
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.gm.cleaner.plugin.home.ui;

import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Locale;

import me.gm.cleaner.plugin.BinderReceiver;
import me.gm.cleaner.plugin.R;
import me.gm.cleaner.plugin.databinding.HomeButtonBinding;
import me.gm.cleaner.plugin.databinding.HomeCardBinding;
import me.gm.cleaner.plugin.databinding.HomeCardButtonBinding;
import me.gm.cleaner.plugin.settings.AppListActivity;

public class HomeAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private static final int TYPE_CARD = View.generateViewId();
    private static final int TYPE_CARD_BUTTON = View.generateViewId();
    private static final int TYPE_BUTTON = View.generateViewId();
    private final HomeActivity activity;

    public HomeAdapter(HomeActivity activity) {
        this.activity = activity;
        BinderReceiver.MODULE_VER.observe(activity, serverVer -> notifyItemRangeChanged(0, 2));
       // ServicePreferences.getInstance().setOnPreferenceChangeListener(() -> notifyItemChanged(1));
    }

    @Override
    public int getItemViewType(int position) {
        switch (position) {
            case 0:
                return TYPE_CARD;
            case 1:
                return TYPE_CARD_BUTTON;
            default:
                return TYPE_BUTTON;
        }
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == TYPE_CARD) {
            HomeCardBinding cardBinding
                    = HomeCardBinding.inflate(LayoutInflater.from(parent.getContext()));
            return new CardHolder(cardBinding);
        } else if (viewType == TYPE_CARD_BUTTON) {
            HomeCardButtonBinding cardButtonBinding
                    = HomeCardButtonBinding.inflate(LayoutInflater.from(parent.getContext()));
            return new CardButtonHolder(cardButtonBinding);
        } else if (viewType == TYPE_BUTTON) {
            HomeButtonBinding buttonBinding
                    = HomeButtonBinding.inflate(LayoutInflater.from(parent.getContext()));
            return new ButtonHolder(buttonBinding);
        } else throw new IllegalArgumentException("undefined view type");
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        int ver = CleanerClient.getServerVersion();
        switch (position) {
            case 0:
                HomeCardBinding cardBinding = ((CardHolder) holder).binding;
                if (showIfNeed(ver != -1, cardBinding.getRoot())) {
                    cardBinding.summary.setText(String.format(activity.getString(R.string.module_version), ver));
                }
                break;
            case 1:
                HomeCardButtonBinding cardButtonBinding = ((CardButtonHolder) holder).binding;
                if (showIfNeed(ver != -1, cardButtonBinding.getRoot())) {
                 //   cardButtonBinding.summary.setText(String.format(Locale.ENGLISH,
                 //           activity.getString(R.string.enabled_app_count), count));
                    cardButtonBinding.getRoot().setOnClickListener(v -> activity.startActivity(
                            new Intent(activity, AppListActivity.class)));
                }
                break;
            case 2:

                break;
            case 3:
                HomeButtonBinding binding = ((ButtonHolder) holder).binding;
                binding.getRoot().setOnClickListener(  v -> {
                    Intent intent = new Intent();
                    intent.setAction("android.intent.action.VIEW");
                    intent.setData(Uri.parse("https://github.com/GuhDoy/Media-Provider-Manager"));
                    activity.startActivity(intent);
                });
                break;
            case 4:
                HomeButtonBinding binding = ((ButtonHolder) holder).binding;
                binding.getRoot().setOnClickListener( v -> {
                    Intent intent = new Intent();
                    intent.setAction("android.intent.action.VIEW");
                    intent.setData(Uri.parse("https://t.me/TabSwitch"));
                    activity.startActivity(intent);
                });
                break;
            case 5:
                HomeButtonBinding binding = ((ButtonHolder) holder).binding;
             binding.getRoot().setOnClickListener(view -> new AlertDialog.Builder(activity)
                     .setMessage("About page is under development.")
                     .setPositiveButton(android.R.string.ok, null)
                     .show());
                break;
        }
    }

    private boolean showIfNeed(boolean isShow, ViewGroup root) {
        ViewGroup.LayoutParams lp;
        if (isShow) {
            lp = new ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        } else {
            lp = new ViewGroup.LayoutParams(0, 0);
        }
        root.setLayoutParams(lp);
        return isShow;
    }

    @Override
    public int getItemCount() {
        return 6;
    }

    public static class CardHolder extends RecyclerView.ViewHolder {
        public HomeCardBinding binding;

        public CardHolder(@NonNull HomeCardBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    public static class CardButtonHolder extends RecyclerView.ViewHolder {
        public HomeCardButtonBinding binding;

        public CardButtonHolder(@NonNull HomeCardButtonBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    public static class ButtonHolder extends RecyclerView.ViewHolder {
        public HomeButtonBinding binding;

        public ButtonHolder(@NonNull HomeButtonBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/me/gm/cleaner/plugin/home/HomeAdapter.kt
package me.gm.cleaner.plugin.home

import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import me.gm.cleaner.plugin.BinderReceiver
import me.gm.cleaner.plugin.BuildConfig
import me.gm.cleaner.plugin.R
import me.gm.cleaner.plugin.dao.ModulePreferences
import me.gm.cleaner.plugin.databinding.HomeButtonBinding
import me.gm.cleaner.plugin.databinding.HomeCardBinding
import me.gm.cleaner.plugin.databinding.HomeCardButtonBinding
import me.gm.cleaner.plugin.settings.AppListActivity
import me.gm.cleaner.plugin.test.TestActivity
import java.util.*

class HomeAdapter(private val activity: HomeActivity) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun getItemViewType(position: Int): Int = when (position) {
        0 -> TYPE_CARD
        1 -> TYPE_CARD_BUTTON
        else -> TYPE_BUTTON
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
        when (viewType) {
            TYPE_CARD -> CardHolder(HomeCardBinding.inflate(LayoutInflater.from(parent.context)))
            TYPE_CARD_BUTTON -> CardButtonHolder(
                HomeCardButtonBinding.inflate(LayoutInflater.from(parent.context))
            )
            TYPE_BUTTON -> ButtonHolder(HomeButtonBinding.inflate(LayoutInflater.from(parent.context)))
            else -> throw IllegalArgumentException("undefined view type")
        }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val ver = BinderReceiver.serviceVersion
        when (position) {
            0 -> {
                val binding = (holder as CardHolder).binding
                when {
                    ver == -1 -> {
                        binding.status.setCardBackgroundColor(activity.getColor(R.color.service_hint_error))
                        binding.icon.setImageResource(R.drawable.ic_baseline_error_24)
                        binding.title.setText(R.string.module_not_active)
                        binding.summary.visibility = View.GONE
                    }
                    BuildConfig.VERSION_CODE != ver -> {
                        binding.status.setCardBackgroundColor(activity.getColor(R.color.service_hint_warn))
                        binding.title.setText(R.string.restart_system)
                        binding.summary.text =
                            String.format(activity.getString(R.string.module_version), ver)
                    }
                    else -> {
                        binding.summary.text =
                            String.format(activity.getString(R.string.module_version), ver)
                    }
                }
                return
            }
            1 -> {
                val binding = (holder as CardButtonHolder).binding
                val count = 114514
                binding.summary.text = String.format(
                    Locale.ENGLISH, activity.getString(R.string.enabled_app_count), count
                )
                binding.root.setOnClickListener {
                    activity.startActivity(Intent(activity, AppListActivity::class.java))
                }
                return
            }
        }
        val binding = (holder as ButtonHolder).binding
        when (position - OFFSET) {
            0 -> {
                binding.icon.setImageResource(R.drawable.ic_outline_history_24)
                binding.title.setText(R.string.usage_record)
                binding.background.setOnClickListener {
                    // activity.startActivity(Intent(activity, ServiceSettingsActivity::class.java))
                }
            }
            1 -> {
                binding.icon.setImageResource(R.drawable.ic_outline_insert_photo_24)
                binding.title.setText(R.string.test)
                binding.background.setOnClickListener {
                    activity.startActivity(Intent(activity, TestActivity::class.java))
                }
            }
            2 -> {
                binding.icon.setImageResource(R.drawable.ic_outline_info_24)
                binding.title.setText(R.string.about)
                binding.background.setOnClickListener {
                    // "https://github.com/GuhDoy/Media-Provider-Manager"
                    // activity.startActivity(Intent(activity, ServiceSettingsActivity::class.java))
                }
            }
        }
    }

    override fun getItemCount(): Int = 3 + OFFSET

    class CardHolder(val binding: HomeCardBinding) : RecyclerView.ViewHolder(binding.root)

    class CardButtonHolder(val binding: HomeCardButtonBinding) :
        RecyclerView.ViewHolder(binding.root)

    class ButtonHolder(val binding: HomeButtonBinding) : RecyclerView.ViewHolder(binding.root)

    companion object {
        const val OFFSET = 2
        private val TYPE_CARD = View.generateViewId()
        private val TYPE_CARD_BUTTON = View.generateViewId()
        private val TYPE_BUTTON = View.generateViewId()
    }

    init {
        BinderReceiver.MODULE_VER.observe(activity) {
            notifyItemRangeChanged(0, 2)
        }
        ModulePreferences.setOnPreferenceChangeListener(object :
            ModulePreferences.PreferencesChangeListener {
            override fun onPreferencesChanged(shouldNotifyservice: Boolean) {
                notifyItemChanged(1)
            }
        })
    }
}
